From: Ilija Tovilo <ilija.tovilo@me.com>
Date: Sun, 3 May 2026 19:57:16 +0200
Subject: GHSA-m33r-qmcv-p97q: [soap] Fix use-after-free after header parsing
 failure with SOAP_PERSISTENCE_SESSION

Fixes GHSA-m33r-qmcv-p97q
Fixes CVE-2026-7261

(cherry picked from commit db2a7f9348fd5dda5fd162061786a664c417bf5b)
(cherry picked from commit 5dd8dd8493d49bb6fcd810a6e9d2ffb6fdc15714)
(cherry picked from commit 63cf032e9675d7d2bbc007c8c787597187a7567b)
(cherry picked from commit dd14d36e31dd99b7589f917924840fe4f46ca022)
(cherry picked from commit 7b354983a33c314b76c594c9c5b790e3b073dcf1)

adapt test for 7.2

(cherry picked from commit f91bcf961ac15eacabf33f86f62c17dbec4a39ab)
(cherry picked from commit ab6fa685773d4efea4de2df4956c97ffd65637e2)
(cherry picked from commit f91ab4e04bc2f254ea1e49e1b76ff55adbbe3892)

adapt test for 5.6
Origin: https://packages.sury.org/php/pool/main/p/php5.6/ (php5.6 5.6.40-104, 0253-GHSA-m33r-qmcv-p97q-soap-Fix-use-after-free-after-he.patch)
Note: refreshed for jessie ELTS php5 5.6.40+dfsg-0+deb8u23, NEWS hunk dropped
---
--- a/ext/soap/soap.c
+++ b/ext/soap/soap.c
@@ -1837,7 +1837,17 @@
 					php_output_discard(TSRMLS_C);
 					soap_server_fault_ex(function, &h->retval, h TSRMLS_CC);
 					efree(fn_name);
-					if (service->type == SOAP_CLASS && soap_obj) {zval_ptr_dtor(&soap_obj);}
+					if (service->type == SOAP_CLASS && soap_obj) {
+#if HAVE_PHP_SESSION && !defined(COMPILE_DL_SESSION)
+					    if (service->soap_class.persistance != SOAP_PERSISTENCE_SESSION) {
+						zval_ptr_dtor(&soap_obj);
+						soap_obj = NULL;
+					    }
+#else
+					    zval_ptr_dtor(&soap_obj);
+					    soap_obj = NULL;
+#endif
+					}
 					goto fail;
 				} else if (EG(exception)) {
 					php_output_discard(TSRMLS_C);
@@ -1852,7 +1862,17 @@
 						soap_server_fault_ex(function, EG(exception), h TSRMLS_CC);
 					}
 					efree(fn_name);
-					if (service->type == SOAP_CLASS && soap_obj) {zval_ptr_dtor(&soap_obj);}
+					if (service->type == SOAP_CLASS && soap_obj) {
+#if HAVE_PHP_SESSION && !defined(COMPILE_DL_SESSION)
+					    if (service->soap_class.persistance != SOAP_PERSISTENCE_SESSION) {
+						zval_ptr_dtor(&soap_obj);
+						soap_obj = NULL;
+					    }
+#else
+					    zval_ptr_dtor(&soap_obj);
+					    soap_obj = NULL;
+#endif
+					}
 					goto fail;
 				}
 			} else if (h->mustUnderstand) {
@@ -2838,7 +2858,7 @@
 	} zend_catch {
 		_bailout = 1;
 	} zend_end_try();
-	
+
 	if (SOAP_GLOBAL(encoding) != NULL) {
 		xmlCharEncCloseFunc(SOAP_GLOBAL(encoding));
 	}
--- a/ext/soap/tests/GHSA-m33r-qmcv-p97q.phpt
+++ b/ext/soap/tests/GHSA-m33r-qmcv-p97q.phpt
@@ -0,0 +1,62 @@
+--TEST--
+GHSA-m33r-qmcv-p97q: Use-after-free after header parsing failure with SOAP_PERSISTENCE_SESSION
+--CREDITS--
+Ilia Alshanetsky (iliaal)
+--EXTENSIONS--
+soap
+session
+--FILE--
+<?php
+
+class Handler {
+  function __call($name, $args) {
+    if ($name == "return") {
+        return new SoapFault('Server', 'denied');
+    } else if ($name == "throw") {
+        throw new SoapFault('Server', 'denied');
+    } else if ($name == "hello") {
+        return 'ok';
+    } else {
+      return new SoapFault("Server","Function $name doesn't exist");
+    }
+  }
+}
+
+session_start();
+
+$srv = new SoapServer(null, ['uri' => 'urn:a']);
+$srv->setClass(Handler::class);
+$srv->setPersistence(SOAP_PERSISTENCE_SESSION);
+
+$x = <<<XML
+<?xml version="1.0" encoding="UTF-8"?>
+<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:a="urn:a">
+    <soap:Header>
+        <a:return/>
+    </soap:Header>
+    <soap:Body>
+        <a:hello/>
+    </soap:Body>
+</soap:Envelope>
+XML;
+$srv->handle($x);
+
+$x = <<<XML
+<?xml version="1.0" encoding="UTF-8"?>
+<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:a="urn:a">
+    <soap:Header>
+        <a:throw/>
+    </soap:Header>
+    <soap:Body>
+        <a:hello/>
+    </soap:Body>
+</soap:Envelope>
+XML;
+$srv->handle($x);
+
+?>
+--EXPECT--
+<?xml version="1.0" encoding="UTF-8"?>
+<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"><SOAP-ENV:Body><SOAP-ENV:Fault><faultcode>SOAP-ENV:Server</faultcode><faultstring>denied</faultstring></SOAP-ENV:Fault></SOAP-ENV:Body></SOAP-ENV:Envelope>
+<?xml version="1.0" encoding="UTF-8"?>
+<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"><SOAP-ENV:Body><SOAP-ENV:Fault><faultcode>SOAP-ENV:Server</faultcode><faultstring>denied</faultstring></SOAP-ENV:Fault></SOAP-ENV:Body></SOAP-ENV:Envelope>
